### PR TITLE
Avoid Issues with GraphReader::GetBeginNodeId

### PR DESCRIPTION
**Pull request by kevinkreiser**
this method, like all the other methods takes a graphtile pointer so that it can speed up lookups if the tile is (likely) already loaded. The problem is it takes this as a reference and then may mutate it to an adjacent tile if the edge in question leaves the tile. The problem with this is that its unintuitive because the begin node which you get back, by definition should be in the same tile as the edge you passed in. There should then, be no need to mutate the tile if it was already the right one. An argument could be made that we want to set it in case the person passed in an uninitialized one so. Perhaps its best to make it do this, especially since this would match all the rest of the functions that work like this. i'll do that

### CODE REVIEW

**Comment by kevinkreiser** [comment]
seems timedep tests are failing. with any lucky they were relying on faulty logic and need to be updated :smile:

File: valhalla/baldr/graphreader.h [hunk 1]
```diff
@@ -637,16 +637,19 @@ class GraphReader {
   /**
    * Method to get the begin node of an edge by using its opposing edges end node
    * @param edge    the edge whose begin node you want
-   * @param tile    reference to a pointer to a const tile
+   * @param tile    reference to a pointer to a const tile containing the begin node
    * @return        returns GraphId of begin node of the edge (empty if couldn't find)
    */
   GraphId GetBeginNodeId(const DirectedEdge* edge, graph_tile_ptr& tile) {
```
Comment (danpat) [inline valhalla/baldr/graphreader.h:643]: If this isn't actually going to mutate the tile pointer any mroe, the signature should be changed `const graph_tile_ptr&`

Comment (danpat) [reply valhalla/baldr/graphreader.h:643]: Oh I see, I missed the added call to `GetGraphTile(opp_edge->endnode(), tile);`.

**Review by nilsnolde — APPROVED** [review]

**Comment by danpat** [comment]
@kevinkreiser Some tests that verify the behaviour of this (and other similar graphptr-mutating functions) might be a good idea.  This is probably the 10th time I've seen a similar issue related to using the wrong tile in tile-boundary situations, it's a bit of a chronic issue.

File: valhalla/baldr/graphreader.h [hunk 2]
```diff
@@ -637,16 +637,19 @@ class GraphReader {
...
-    if (!GetGraphTile(edge->endnode(), tile))
+    // grab the end node maybe in an adjacent tile
+    graph_tile_ptr maybe_other_tile = tile;
+    if (!GetGraphTile(edge->endnode(), maybe_other_tile))
       return {};
-    const auto* node = tile->node(edge->endnode());
-    // grab the opp edges end node
-    const auto* opp_edge = tile->directededge(node->edge_index() + edge->opp_index());
+    const auto* node = maybe_other_tile->node(edge->endnode());
+    // grab the opp edge also could be in this adjacent tile
```
Comment (julianskartor) [inline valhalla/baldr/graphreader.h:649]: I was wondering if we could write out "opposite" in the comment? Also maybe some punctuation is missing in the sentence? :thinking:

Comment (nilsnolde) [reply valhalla/baldr/graphreader.h:649]: maybe a comma missing, but "opp edge" should be readily understandable for someone coming this far in the code base. it's literally all over the comments & var names, can't avoid it

**Pull request merged**